tiledb/cloud/client.py

```diff
@@ -375,148 +372,154 @@ def find_organization_or_user_for_default_charges(user):
-                "GET",
-                "PUT",
-                "DELETE",
-                "OPTIONS",
-                "TRACE",
-                "POST",
-                "PATCH",
-            ],
-            raise_on_status=False,
-            # Don't remove any headers on redirect
-            remove_headers_on_redirect=[],
-        )
-        self.update_clients()
-
-    def retry_mode(self, mode="default"):
-        """
-
-        :param mode: Retry mode ["default", "forceful", "disabled"]
-        :return:
-        """
-        if mode == "default":
-            self.set_default_retries()
-        elif mode == "forceful":
-            self.set_forceful_retries()
-        elif mode == "disabled":
-            self.set_disable_retries()
-        else:
-            raise Exception(
-                f"unsupported retry mode {mode!r}. Valid options are default, forceful or disabled"
-            )
-
-        self.update_clients()
-
-    def set_threads(self, threads=os.cpu_count() * 2):
-        """
-        Update thread pool sizes for async functionality
-        :param threads:
-        :return:
-        """
-        self.pool_threads = threads
-        self.update_clients()
-
-    def __get_array_api(self):
-        return rest_api.ArrayApi(
-            rest_api.ApiClient(
-                configuration=config.config, pool_threads=self.pool_threads
-            )
-        )
-
-    def __get_user_api(self):
-        return rest_api.UserApi(
-            rest_api.ApiClient(
-                configuration=config.config, pool_threads=self.pool_threads
-            )
-        )
-
-    def __get_organization_api(self):
-        return rest_api.OrganizationApi(
-            rest_api.ApiClient(
-                configuration=config.config, pool_threads=self.pool_threads
-            )
-        )
-
-    def __get_udf_api(self):
-        return rest_api.UdfApi(
-            rest_api.ApiClient(
-                configuration=config.config, pool_threads=self.pool_threads
-            )
-        )
-
-    def __get_tasks_api(self):
-        return rest_api.TasksApi(
-            rest_api.ApiClient(
-                configuration=config.config, pool_threads=self.pool_threads
-            )
-        )
-
-    def __get_sql_api(self):
-        return rest_api.SqlApi(
-            rest_api.ApiClient(
-                configuration=config.config, pool_threads=self.pool_threads
-            )
-        )
-
-    def __get_notebook_api(self):
-        return rest_api.NotebookApi(
-            rest_api.ApiClient(
-                configuration=config.config, pool_threads=self.pool_threads
-            )
-        )
+        self.retry_mode(RetryMode.FORCEFUL)
+
+    def retry_mode(self, mode: RetryOrStr = RetryMode.DEFAULT):
+        """Sets how we should retry requests and updates API instances."""
+        mode = RetryMode.maybe_from(mode)
+        config.config.retries = _RETRY_CONFIGS[mode]
+        client = rest_api.ApiClient(config.config)
```

Comment:
Is this going to initialize another (multiprocessing) ThreadPool? https://github.com/TileDB-Inc/TileDB-Cloud-Py/blob/641edd135d7c59c40544a1e8826acfd44e2fe392/tiledb/cloud/rest_api/api_client.py#L112

or is that code lazy and never-called?

Reply:
It is, yes, but:

- The threads are only ever actually started when a request is made, so it is lazy in that sense. Since we never use the async functionality built into the API client, they will never be started.
- The previous code created one thread pool for every individual API instance, so we still end up creating fewer never-used thread pools.

Reply:
👍 thanks